jfr/src/main/java/io/netty/jfr/JfrChannelHandler.java

```diff
@@ -0,0 +1,417 @@
+            channelWritabilityChangedEvent.setChannel(ctx.channel());
+            channelWritabilityChangedEvent.setIsWritable(ctx.channel().isWritable());
+            channelWritabilityChangedEvent.commit();
+        }
+    }
+
+    @Override
+    public void flush(ChannelHandlerContext ctx) {
+        final FlushEvent flushEvent = new FlushEvent();
+        flushEvent.begin();
+        ctx.flush();
+        flushEvent.end();
+        if (flushEvent.shouldCommit()) {
+            flushEvent.setChannel(ctx.channel());
+            flushEvent.commit();
+        }
+    }
+
+    // Event types
+
+    @Category({ "Netty", "Channel" })
+    @Enabled(true)
+    @Threshold("1 us")
+    private abstract static class ChannelEvent extends Event {
+        @Label("Channel")
+        String channel;
+        @Label("Long Channel ID")
+        String longChannelId;
+        @Label("Short Channel ID")
+        String shortChannelId;
+
+        void setChannel(Channel channel) {
+            this.channel = String.valueOf(channel);
+            //TODO: More specific channel fields?
+            this.longChannelId = channel.id().asLongText();
+            this.shortChannelId = channel.id().asShortText();
+        }
+    }
+
+    @Threshold("20 ms")
+    private abstract static class ChannelIOEvent extends ChannelEvent {
+    }
+
+    @Label("Bind")
+    @Name("io.netty.channel.Bind")
+    private static final class BindEvent extends ChannelEvent {
+        @Label("Local Address")
+        String localAddress;
+
+        void setLocalAddress(SocketAddress localAddress) {
+            this.localAddress = String.valueOf(localAddress);
+        }
+    }
+
+    @Label("Write")
+    @Name("io.netty.channel.Write")
+    private static final class WriteEvent extends ChannelIOEvent {
+        @Label("Message")
+        String msg;
+
+        void setMsg(Object msg) {
+            this.msg = String.valueOf(msg);
+        }
+    }
+
+    @Label("Channel Read")
+    @Name("io.netty.channel.ChannelRead")
+    private static final class ChannelReadEvent extends ChannelIOEvent {
+        @Label("Message")
+        String msg;
+
+        void setMsg(Object msg) {
+            this.msg = String.valueOf(msg);
+        }
+    }
+
+    @Label("Channel Read Complete")
+    @Name("io.netty.channel.ChannelReadComplete")
+    private static final class ChannelReadCompleteEvent extends ChannelIOEvent {
+    }
+
+    @Label("Close")
+    @Name("io.netty.channel.Close")
+    private static final class CloseEvent extends ChannelEvent {
+    }
+
+    @Label("Disconnect")
+    @Name("io.netty.channel.Disconnect")
+    private static final class DisconnectEvent extends ChannelEvent {
+    }
+
+    @Label("Connect")
+    @Name("io.netty.channel.Connect")
+    private static final class ConnectEvent extends ChannelEvent {
+        @Label("Remote Address")
+        String remoteAddress;
+        @Label("Local Address")
+        String localAddress;
```

Comment:
In my opinion I think it would be interesting to split the address from the port.
This could ease analysis in JMC I think.